src/Symfony/Component/Routing/Generator/UrlGenerator.php

```diff
@@ -266,20 +274,42 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
         $fragment = isset($extra['_fragment']) ? $extra['_fragment'] : '';
         unset($extra['_fragment']);
 
-        if ($extra && $query = http_build_query($extra, '', '&')) {
-            // "/" and "?" can be left decoded for better user experience, see
+        // ignore null values
+        $extra = array_filter($extra, function($value) { return isset($value); });
+
+        if ($extra) {
+            if ($this->charset != 'UTF-8') {
```

Comment:
1. `Always use identical comparison unless you need type juggling;`
2. `Use Yoda conditions when checking a variable against an expression to avoid an accidental assignment inside the condition statement (this applies to ==, !=, ===, and !==);`

https://symfony.com/doc/current/contributing/code/standards.html#structure

`if ('UTF-8' !== $this->charset) {`